### ISSUE TITLE
Inheritance / Polymorphism

### ISSUE THREAD

**Issue by cmelchior** [realm/realm-java#761]
A highly request feature is polymorphism. Mostly for use in ListAdapters, but a lot of other scenarios exists.

We need support for the following:
- Abstract RealmObject classes
- Query support

Example
```
public abstract Animal extends RealmObject {
  private String name;
}

public class Dog extends Animal {
  private boolean willPlayCatch;
}

public class Spider extends Animal {
  private boolean isScary;
}

// Following should be possible
RealmResults<Animal> animals = realm.where(Animal.class).equalsTo("name","Foo").findAll();
String name = animals.get(0).getName();

RealmResults<Dog> dogs = realm.where(Dog.class).equalsTo("name","Foo").and().equalsTo("willPlayCatch", true).findAll();
String name = dogs.get(0).getName();
```

**Comment by T-Spoon** [comment]
Any estimate on a time frame for this?  I'm hoping to use Realm in an upcoming project, but without support for Model inheritance I'll have to hold off on trying it out.

Looking forward to where you guys take this project though!

**Comment by cmelchior** [comment]
Hi @T-Spoon 
We have a couple of other major features we would like to look at first, so we don't have a timeframe yet, sorry. If inheritance is a dealbreaker for you, you should probably not use us this time around. But we do hope you will come back :)

**Comment by caroth92** [comment]
:+1:

**Comment by msegers** [comment]
Still no estimates? Going to write a lot of duplicate code since I want to use Realm anyways. Works fine in swift already. Monitoring this.

**Comment by cmelchior** [comment]
Unfortunately not. The support for this in Objective C and Swift is actually just sharing fields. It is not proper polymorphism. This is good enough for some use cases, but we feel that without being able to query on abstract super classes it will be less useful on Android. Right now we are focusing on migrations, null support and async querying, but hopefully we will be able to tackle this afterwards.

**Comment by nolanamy** [comment]
So is there any way to share fields on Android? We have 5 models that all share the same synchronization-related fields and code. We use inheritance in our current SQLite models; if we switch to Realm, will we have to duplicate the synchronization fields across each of the 5 model classes?

As a workaround, I'm thinking about having those classes implement an interface with getters and setters for the shared fields, which at least let me share the synchronization code... is there a better way?

**Comment by AliSoftware** [comment]
While waiting for this inheritance feature to be available, is there any workaround, like a good practice to design our RLMObjects and model to make such use cases possible?

My need is with a typical object model having, say, Animals objets, subclassed as Dogs and Cats, and I want sometimes to fetch all the Animals (to display in a List only common properties of those animals like their name), and sometimes (e.g. when I tap on an item in the list to show the details) to access properties of the specific class Dog or Cat.

Given that I can't use inheritance to represent that in Realm yet, how would you guys represent your data model instead to make this use-case possible? Adding a relationship from the Dog and Cat RLMObjects to the Animal RLMObject (composition instead of inheritance) maybe?

**Comment by emanuelez** [comment]
The current solution is to use composition instead of inheritance. 
There are lots of good argument about it (Gang of Four and Joshua Bloch to mention some prestigious supporters of this approach).

On the other hand we are considering to allow inheritance, but it would not allow queries. 
For example: Animal which is extended both by Dog, Cat and Duck. You would not be able to query Animals with for legs and have all Dogs and Cats, but not Ducks. We feel this would be a very crippling behavior but are eager to listen to more opinions.

**Comment by nolanamy** [comment]
@emanuelez Is @AliSoftware's composition-over-inheritance example what you have in mind?

**Comment by rovkinmax** [comment]
See other model: Im have base object     
``` java
public abstract class BaseObject {
    private String guid;

    private Map<String, String> name;

    private List<String> tags;

    private double latitude;

    private double longitude;

    private LatLng position;

    @Nullable
    public String getName(String locale) {
        if (name != null) {
            return name.get(locale);
        }
        return null;
    }

    public String getName() {
        return getName(getDefaultLocale());
    }

}
```

and some other child
``` java
public class City extends BaseObject {

    int version;

    boolean current = false;
}
```

```java
public class Station extends BaseObject {
   
    Map<String, String> next;

    List<String> routes;

    private List<String> bus;
    private List<String> shuttle;
    private List<String> tram;
    private List<String> troll;
}
````

```java
public class RoutePoint extends BaseObject {
    int direction;
    String routeGuid;
}
```
and few other

How much do I need to copy paste the code and how much it will generate errors?

**Comment by StErMi** [comment]
Hi, is there any update about this issue?

**Comment by cmelchior** [comment]
Hi @StErMi We are currently focusing on finishing other features, and we cannot move forward with this until we made changes to the underlying storage engine as it must support querying across abstract types. At this time I am afraid I don't have any timeline.

**Comment by srikalyan** [comment]
@cmelchior, I can help you guys building this feature but I need little help from you guys in order to get starting so if you can guide me with the kick off process that would be great.

**Comment by claudioredi** [comment]
I appreciate your hard work on this excellent piece of software but it's a big pain in the ass not being able to use inheritance as we do in swift. I have a big model in swift full of inheritance that I have to implement in Android and now I realize this is not allowed...I never imagined such a restriction could exist on a object oriented language component.

IMO this is a very important thing but this ticket was opened for almost a year without any estimated timeline so I assume the solution is not near :)

**Comment by cmelchior** [comment]
Unfortunately we don't have a timeline yet. We have a long list of features we would like to do, and we are continuously evaluating the order we do them in, so feedback such as yours is much appreciated. 

Note that Swift offers inheritance but not polymorphism, which we think would be the expectation for any Java developer for this feature. So for now we recommend using composition instead of inheritance.

**Comment by cmelchior** [comment]
@srikalyan Sorry for the late reply. Thank you for your offer, but this feature unfortunately require changes to the underlying core engine (all issues marked "waiting-for-core" does that), but if you find any issues without that label, we would love to start a discussion on them if you want to help.

**Comment by mlusas** [comment]
@cmelchior Just echoing the desire to have polymorphic relationships supported in Realm.  Also, some polymorphic solutions seem to be Integer focused; it would be good to have something that supports Binary and String/Hex as well.

Until Realm releases polymorphic support, we're building workarounds at the cost of performance, code efficiency, and model simplicity.

**Comment by vergauwen-simon** [comment]
> IMO this is a very important thing but this ticket was opened for almost a year without any estimated timeline so I assume the solution is not near :)

I've used Realm for quite a few projects. And having inheritance would save me a lot of trouble. Like someone said before, this ticket was opened a year ago. Realm is my fav library out their for database related stuff in Android. And I hope this feature will be added soon, I think Realm **could** be a big player in how DB related things get handled in Android. It is great as it is, but missing inheritance in a OOPL like java I think it is a major let down for a lot of people.

I assume there is still no estimated timeline?

Keep up the good work though!

**Comment by rpip** [comment]
:+1:

**Comment by savekirk** [comment]
+1

**Comment by borwoj** [comment]
+1 Hopefully this feature is still on the table

**Comment by dbauduin** [comment]
+1000

**Comment by tuksiarz** [comment]
+1! :)

**Comment by monossido** [comment]
+1 plz

**Comment by eyecreate** [comment]
+1

**Comment by yosrah** [comment]
+1

**Comment by quangson91** [comment]
+1

**Comment by mformetal** [comment]
+1

**Comment by kmahmood74** [comment]
+1

**Comment by renegade037** [comment]
+1

**Comment by drcw** [comment]
Yeah, we're all waiting! Its a very cool DB and we all want INHERITANCE! :D

**Comment by joaoBeno** [comment]
@emanuelez and @cmelchior I think this proposed approach is acceptable, since the main drawback of not having this is having to repeat a lot of code for objects that share the same behavior. It should be noted on queries that we can't query objects based on a common ancestor, but we should be able to make joint queries (Like: find `Dog` and `Cat` where color is black)... On the future, if the demand for ancestor querying grows, then we can think on some way to create a relation between the parent object and it's siblings to allow the query mentioned by Emanuele...

**Comment by wpx121314** [comment]
+1 wait for

**Comment by rodrigophn7** [comment]
+1 waiting...

**Comment by chmoder** [comment]
+1

**Comment by samyakjain** [comment]
+1

**Comment by leasual** [comment]
+1

**Comment by VenomVendor** [comment]
Tracking

**Comment by bhavyakilari** [comment]
+1

**Comment by juanpinto** [comment]
+1

**Comment by tonyshkurenko** [comment]
One year passed away

**Comment by baratsemet** [comment]
@cmelchior It's been a year and more, still have NO answer for this?

**Comment by bmunkholm** [comment]
Wauu - yeah time flies quickly. We would really love to tackle this very soon. During the past year we have closed about 900 issues. This hasn't been one of them since we have always prioritized supporting/helping developers and fixing bugs as 1. priority. Besides that our priority is to fix things that doesn't have any workarounds. We have also gone completely transparent with our issue mgmt and priorities, and you can see a more detailed priority list if you view the issues in waffle.io.
Our "S:P1 backlog" list will be tackled first and we will make a 1.0 release. After that we expect this to be among the top 3 features to support. This doesn't give you a date, but we really can't predict it - it depends on the future support load. But I hope it gives at least some insight into what we are thinking.

**Comment by lectricas** [comment]
+1

**Comment by Drakuwa** [comment]
Oh come on people, do not just comment "+1", I know my comment also has no point, but, you could all just subscribe to the thread and wait for an actual valid response from someone from the team. This way you're not just flooding the issue with unnecessary entries...

**Comment by lectricas** [comment]
that was the most simple way to show that we are all waiting for this feature.

**Comment by beeender** [comment]
Comments from #2643 RealmObject should also be able to extend other non-RealmObject/abstract object.

**Comment by luisribeiro** [comment]
Hi there. Do you know when we will be able to extend a class that extends RealmObject or implements RealmModel?

**Comment by NitroG42** [comment]
Hi, I want to try composition over inheritance for now. Let's say we do what @emanuelez suggested : 
```
class Animal {
int numberOfLegs 
String name
}
```
and 
```
class Dog:  { 
Int furSize
Animal animal
}
```
```
class Duck:  { 
Int featherSize
Animal animal
}
```

Now if I query for all Animal.class, I will get all of them (that is awesome btw) **BUT**, for a given Animal, I can't get the related Dog or Duck (at least i don't know how !). 
Do you have any suggestion @emanuelez ?

**Comment by emanuelez** [comment]
@NitroG42 If you add a unique id field to the animal class, then you can query for it for all you dogs and ducks. For a cleaner approach you will have to wait for backlinks (https://github.com/realm/realm-java/issues/607)

**Comment by uguraykilic** [comment]
+1

**Comment by henriquedesousa** [comment]
I'm thumbing this down, because after one year, there's not even an example code on we can overcome this.

**Comment by henriquedesousa** [comment]
We've tried composition, but introduced way more problems. Now, instead of developing, we're lots of time hacking the models to fit our bussiness rules. Mangle it with retrofit, deserializers, and all hell broke loose!

**Comment by valokafor** [comment]
This is super painful, I cannot forgive myself for making a huge assumption that Realm supports inheritance. I have a project with a fast-approaching deadline and I had ripped out the good old SQLite/ContentProvider combo and implemented Realm for faster performance and now I hit the wall with Inheritance and composition is not an option for me. I will keep tuned in for this feature. 

Overall, I am very happy with all that Realm has offered and I wish the team luck with finding a solution for this.

**Comment by akiander** [comment]
I'm trying to use Realm and I just ran into this limitation. Is there a timeline for support objects that already have their own inheritance? Composition is not an option for us right now. Is there a timeline for this yet?

**Comment by yuriy-larin** [comment]
+1 - I vote for it. More then a year has passed.

**Comment by schatekar** [comment]
I had just started migrating from ActiveAndroid to realm and realised that subclassing is still not supported. I think I will have to go back to looking for another ORM.

**Comment by yuriy-larin** [comment]
I also switched back to SQLite for now as I need to work with database from Service and Activity code.
I found it hard to use realm in such 'multi-threaded' environment.

**Comment by bryant1410** [comment]
I can recommend [DbFlow](https://github.com/Raizlabs/DBFlow), which supports polymorphism

**Comment by schatekar** [comment]
@bryant1410 Had a quick look at DbFlow. I did not like the code it auto-generates. This is totally a personal preference though. why you think DbFlow is better than something like ActiveAndroid?

**Comment by bryant1410** [comment]
Do you mean that you don't like its API? Or the code itself? If you don't like the auto-generated code, this is not a problem as you don't have to read it very often (if at all).

In my opinion the game-changing factor here is the performance. ActiveAndroid relies heavily on reflection, while DBFlow uses annotation processing in order to generate classes and to avoid using reflection as little as possible. Reflection is usually slow in Java, and not using it has a big impact. You can take a look to [DBFlow's company benchmark about this](https://www.raizlabs.com/dev/2015/02/go-dbflow-fastest-android-orm-database-library/).

**Comment by olivierg13** [comment]
+1 - If you could find a solution for the inheritance to work, that would be awesome!
:)

**Comment by PaulMuriithi** [comment]
I have used Realm for quite a while but lack of support of this fundamental OOP characteristic is quite disappointing, a year and a half is quite a long time to have built the required support.

**Comment by leonapse** [comment]
Just ran into this problem. It would be very great if this feature gets prioritized. Gonna have to abandon Realm for now and use other ones. Updates from devs would definitely be appreciated! <O

**Comment by lordplagus02** [comment]
This issue means I can't attach any JSONAPI deserializers like [moshi-jsonapi](https://github.com/kamikat/moshi-jsonapi) in Android (which requires that you extend your class with `Resource`, which means you can't extend `RealmObject` or vice versa. Anybody have a temporary workaround for something like this?

**Comment by beeender** [comment]
@lordplagus02 See https://realm.io/docs/java/latest/#realmmodel-interface

**Comment by lordplagus02** [comment]
@beender I think you've just saved my day, what with deadlines fast approaching and solutions undiscovered... you know how it is

**Comment by lordplagus02** [comment]
@beeender It appears that it still doesn't actually help, you still can't extend or implement anything that isn't Realm related...

**Comment by beeender** [comment]
@lordplagus02 Sorry I missed that part "moshi-jsonapi in Android (which requires that you extend your class with Resource" ... I think for now you need to find some other ways if you want to use moshi-jsonapi and Realm together, maybe create some delegate class? Not sure if there is an easy way to do it.

**Comment by lordplagus02** [comment]
@beeender easiest way was to revert to moshi-jsonapi v1.x because it doesn't require that you extend `Resource`. I think due to the current limitations of Realm, we're still awhile away from a full jsonapi/realm/syncadapter implementation.

**Comment by Zhuinden** [comment]
Looking at it, a possibility is forking the project and rewriting it to use interface instead of hard-coding their abstract class

**Comment by navjotbedi** [comment]
+1

**Comment by Zhuinden** [comment]
``` java
enum ChildType {
    CAT, DOG
}

class Animal extends RealmObject implements CatInterface, DogInterface {
    // animal attributes
    @PrimaryKey
    @Required
    private String id;

    @Required
    @Index
    private String type;

    public String getType() {
         return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public void setChildType(ChildType childType) {
        setType(childType.name());
    }

    public ChildType getChildType() {
        return ChildType.valueOf(this.type);
    }

    public Animal() {
    }

    // cat attributes
    private String catSomething;
        // getter, setter

    // dog attributes
    private String dogSomething;
        // getter, setter
}


interface CatInterface {
    String getCatSomething();

    void setCatSomething(String catSomething);
}

interface DogInterface {
    String getDogSomething();

    void setDogSomething(String dogSomething);
}

```

Then

``` java
    RealmResults<CatInterface> cats = (RealmRseults<CatInterface>)((RealmResults)(realm.where(Animal.class)
                                                                       .equalTo(AnimalFields.TYPE, ChildType.CAT.name())
                                                                       .findAll()));
```

**Comment by renaudfavier** [comment]
@Zhuinden Nice ! thanks it seems like what I did but better

**Comment by lordplagus02** [comment]
Can I bump this? Any update on this feature?

**Comment by luizfelippe** [comment]
I was just modifying my Realm classes to work with Polymorphism but then I faced this issue ):

**Comment by Zhuinden** [comment]
Only interfaces are supported at the moment, base classes are not.

**Comment by bogomazov** [comment]
I assumed that inheritance is here by default and refactored my code around it without compiling in the process (yes, I am naive but if the support won't be added I am gonna have to have lots of code duplications around primary keys for each model, what the point of using Realm then?)
Waiting for an update

**Comment by Zhuinden** [comment]
@bogomazov you can always use an interface, and a repository implementation for `T extends YourInterface`.

**Comment by lordplagus02** [comment]
That's what I did, but it causes for quite a bit of un-dry code in your object classes. The interface is just useful for your getters and setters at the end of the day. It desperately needs some polymorphism.

**Comment by valokafor** [comment]
@lordplagus02  and @bogomazov  can you show me an example of the Repository implementation that you got working for you. I am using Repository implementation, maybe I am doing it wrong.

My model class was simple, a LineItem that inherits from a Product like so:

```
public class LineItem extends Product {
    private long id;
    private int quantity;

    public LineItem(Product product, int quantity) {
        super(product);
        this.quantity = quantity;        
    }
}
```
I have now changed this to this as a work around.

```
@RealmClass
public class LineItem implements RealmModel {
    private long id;
    private int quantity;
    private long productId;
    private Product product;

    public LineItem() {
    }

    public LineItem(Product product, int quantity) {
        this.quantity = quantity;
        this.productId = product.getId();
        this.product = product;
    }
}
```

Here is my repository interface, how can I change my repository implementation to support polymorphism.

```
 public interface Repository{
        List<LineItem> getAllLineItemsInATransaction(long transactionId);
        long saveLineItem(LineItem lineItem, AsyncQueryListener listener);
        void updateLineItem(LineItem lineItem, AsyncQueryListener listener);
        LineItem getLineItemById(long id);
        void deleteLineItem(long id, AsyncQueryListener listener);
    }
```

**Comment by Zhuinden** [comment]
I do not see a single place where you're actually sharing fields with `Product`.

**Comment by valokafor** [comment]
LineItem is inheriting from Product and then adding properties that are LineItem specific such as quantity, discount, refund, etc. I have not shown those for brevity. The goal is not to duplicate the properties of Product such as name, image, price, etc again in LineItem hence the inheritance.

**Comment by kaptanblack** [comment]
+1

**Comment by lordplagus02** [comment]
*sigh* this is messing with my zen

**Comment by bubakazouba** [comment]
Wow 2 years

**Comment by Zhuinden** [comment]
it is a complicated problem, and using common interfaces is a fairly simple workaround

**Comment by chanakin** [comment]
Can someone help with this? http://stackoverflow.com/questions/42259735/composition-over-inheritance-for-realmobjects-with-gson-serialization

I'm totally stuck. This is incredibly frustrating and no one seems to have posted a thorough example of how to work around this issue, and if I can't figure out a workaround, then Realm isn't even an option. I'm astonished they haven't addressed inheritance in two years; I get that development takes time, but inheritance is one of the basic principles of OOP. How can anyone use this if their app has even a modicum of complexity?

**Comment by Zhuinden** [comment]
You should map your JSON responses into a Realm schema. Don't just directly map GSON's classes into Realm, it's generally a bad idea to assume that your API calls and your database will 1-to-1 match each other directly.

**Comment by wingu-wiktor** [comment]
Why don't you state clearly in RealmModel interface that class implementing it cannot extend any class? Documentation in current form leads to opposite conclusion.

**Comment by Zhuinden** [comment]
I really do think the fact that you can't inherit from a class that only has `@Ignore`'d fields or no fields at all is a bug.

**Comment by beeender** [comment]
@Zhuinden But if an object only has `@Ignore` fields or no fields, what is the point of being a `RealmObject`?

**Comment by Zhuinden** [comment]
@beeender example code

``` java
public class BaseObservable implements Observable {
    @Ignore
    private transient PropertyChangeRegistry mCallbacks;

    @Override
    public synchronized void addOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        if (mCallbacks == null) {
            mCallbacks = new PropertyChangeRegistry();
        }
        mCallbacks.add(callback);
    }

    @Override
    public synchronized void removeOnPropertyChangedCallback(OnPropertyChangedCallback callback) {
        if (mCallbacks != null) {
            mCallbacks.remove(callback);
        }
    }

    @Override
    public synchronized void notifyChange() {
        if (mCallbacks != null) {
            mCallbacks.notifyCallbacks(this, 0, null);
        }
    }

    public void notifyPropertyChanged(int fieldId) {
        if (mCallbacks != null) {
            mCallbacks.notifyCallbacks(this, fieldId, null);
        }
    }
}

@RealmClass
public class ObservableDog extends BaseObservable implements RealmModel {
    @PrimaryKey
    private long id;

    private String text;

    @Bindable
    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
        if(!isManaged()) {
            notifyPropertyChanged(BR.id);
        }
    }

    @Bindable
    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
        if(!isManaged()) {
            notifyPropertyChanged(BR.text);
        }
    }
}
```

This code fails on the Realm annotation processor

**Comment by srinivas3120** [comment]
any update on this issue ?

**Comment by VedavyasBhat** [comment]
Any update on this issue?

I have models that look like this:
```
class Chat extends RealmObject {
    RealmList<TextMessage> textMessages;
    RealmList<ImageMessage> imageMessages;
}
class Message extends RealmObject {
    //some common fields
}
class TextMessage extends RealmObject {
    Message baseMessage;
}
class ImageMessage extends RealmObject {
    Message baseMessage;
}
```
instead of looking something like this:
```
class Chat extends RealmObject {
    RealmList<Message> messages
}
class Message extends RealmObject {
    //some common fields
}
class TextMessage extends Message {}
class ImageMessage extends Message {}

```
and it's a nightmare to add more message types or even write queries.

**Comment by Zhuinden** [comment]
I do not see why you have any more realmObjects beyond `Message`

**Comment by VedavyasBhat** [comment]
@Zhuinden Because I have more fields in `TextMessage` and `ImageMessage`, which I haven't included here for the sake of readability.

**Comment by Zhuinden** [comment]
@VedavyasBhat I stand by my question. It would have been easier to just include nullable fields and a `@Index String type` field.

**Comment by sluedecke** [comment]
@Zhuinden: I see what you have in mind, but I have the same issue as @VedavyasBhat:  I want to be able to attach `Actions` and `Observations` to an object, both being an event in time with some shared attributes.  But since both have >10 unique attributes I would end up with a large class `Event` which I consider hard to maintain.

Also I am afraid that I will loose productivity when using code completion and that it will add a source for errors (accessing fields from the wrong type, ...).

But it is a feasible workaround.

**Comment by Zhuinden** [comment]
When I had this, I prefixed the fields with what event type (in my case, post type) the given field belonged to.

**Comment by Heepie** [comment]
1. What is the problem in this code? I tried to follow above code.
but, 
Error:(11, 8) error: Realm model classes must either extend RealmObject or implement RealmModel to be considered a valid model class

Component.java

```
public abstract class Component extends RealmObject {

    private String name;

    public Component(String name) {
        this.name = name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
}
```

Card.java

```
public class Card extends Component {
    public Card(String name) {
        super(name);
    }
}
```


Folder.java

```
public class Folder extends Component {
    private RealmList<Component> list = new RealmList<Component>();

    public Folder(String name) {
        super(name);
    }

    public Folder add(Component item) {
        list.add(item);
        return this;
    }

    public RealmList<Component> getList() {
        return list;
    }
}
```

2. is there something solution??

**Comment by Zhuinden** [comment]
The only "direct base class" a RealmObject can extend is `extends RealmObject`.

Anything else must be done through interface.

**Comment by Heepie** [comment]
If I want to make the class through realm, I must have to `extends RealmObject`. Right??

So then, 
```
public class Card extends RealmObject implements Cardable {
    ...
}

```
```
public class Folder extends RealmObject implement Folderable {
...
}
```
Right??

And I have a question. is it impossible above example code?? (Animal, Dog, Spider and getting RealmResult through Animal etc)

**Comment by Zhuinden** [comment]
@Heepie to obtain all those types of animals as `Animal`, currently you'd need 

``` java
public class Animal extends RealmObject {
    @PrimaryKey
    private String animalId;

    private String dogName;
    private String spiderPoison;
    /// ...

    @Index
    private String animalType; // DOG, SPIDER, ...
```

Basically a single RealmObject class that merges the types together and its type is determined by a discriminator column ([Single Table Inheritance](https://en.wikipedia.org/wiki/Single_Table_Inheritance)).

**Comment by Heepie** [comment]
@Zhuinden Think you so much :)

I'll try Single Table Inheritance!

**Comment by anuragopenxcell** [comment]
I started with implementing Realm in my project and hit this roadblock on the very first insert:| Do you guys have any eta on this yet?

**Comment by beeender** [comment]
@anuragopenxcell realm-core started some experimental work on this feature, but this still has a low priority for us. We cannot give a estimation on it. sorry.

**Comment by hardysim** [comment]
@beeender to bad that is has such a low priority as it's the most wanted feature / issue here

![image](https://user-images.githubusercontent.com/3737512/27780846-9693869a-5fcb-11e7-8b8e-30e1751a2436.png)

**Comment by Zhuinden** [comment]
Single-Table inheritance still works. If you design the schema taking the constraints of RealmModels in consideration, everything still works just fine. Especially considering you can mimic the behavior with interfaces if you intend to go so far.

**Comment by omidraha** [comment]
This issue [block](https://github.com/davideas/FlexibleAdapter/blob/extensions/flexible-adapter-realm/src/main/java/eu/davidea/flexibleadapter/realm/AbstractRealmFlexibleItem.java#L14) [#166 ](https://github.com/davideas/FlexibleAdapter/issues/166)- Added Realm extensions to [FlexibleAdapter](https://github.com/davideas/FlexibleAdapter).
Any estimate on doing this?

**Comment by Zhuinden** [comment]
@omidraha they didn't say blocked, just inconvenient. (having to implement IFlexible on each RealmObject)

**Comment by YamilMarques** [comment]
Should be nice to have this feature working, and looks like a lot of people are waiting/needing it for more than 2 years. There is any way to increase the priority?

**Comment by Zhuinden** [comment]
@YamilMarques probably after some other essential features that are on the way such as cascade deletes and list of primitives

**Comment by chmoder** [comment]
Thanks for the update @zhuinden.  I've been using realm without the polymorphism feature but wish it had it especially at the beginning of our project.  

While all three of these features are very important, now I think list of primitives should be a step higher priority than polymorphism.

**Comment by dvdmunckhof** [comment]
Thereafter, the error message should be more clear:

>Error:(9, 8) error: Valid model classes must either extend RealmObject or implement RealmModel.

I did implement RealmModel, however I extended another class too. I assumed polymorphism would be supported and the error message doesn't state otherwise.

**Comment by cotfas** [comment]
Any news ?

**Comment by KevinYaoooooo** [comment]
Time flies, it's been 2 years, still no news?

**Comment by Zhuinden** [comment]
Well a bunch of other things were added and are on the table

**Comment by cotfas** [comment]
I`m really interested in this and it would be awesome to have it.

**Comment by thenewmr** [comment]
Backlog label removed.  Does that mean it's being worked on?  Please say yes :)

**Comment by cmelchior** [comment]
We have begun some preliminary work yes (don't hold your breath though, we want to finish #575 first), but it is mostly because we switched to Zenhub to do issue management which has its own way of tracking it instead of GitHub labels.

**Comment by poovamraj** [comment]
+1

**Comment by cotfas** [comment]
+1

**Comment by NirKimchiZabari** [comment]
+1

**Comment by markiantorno** [comment]
If you guys fix issues 761 and 575, I will actually start using it on my projects again. I stopped using Realm because of the limitations these unfinished tickets imposed on architecture. These changes would be incredible.

**Comment by Zhuinden** [comment]
@markiantorno well, #575 is in the works (although I think the "SQLite approach" of storing the primitive list as a single String field, then transform it to List in the getter worked wonders), in the meantime the single-table inheritance approach works well, although I do admit having to do that yourself imposes limitation on how you'd normally portray your object hierarchy.

**Comment by markiantorno** [comment]
@Zhuinden ...especially when you're trying to build support libraries for pre-existing tools with really deep hierarchies (like http://hl7.org/fhir/ for example). Don't get me wrong, I love Realm and think it's fantastic. It just needs these two features for me to really integrate it into the software I'm writing at the hospital.

**Comment by Damnum** [comment]
Please update the docs and mention that it's not possible: https://realm.io/docs/java/latest/#customizing-objects

**Comment by hhhung** [comment]
Please make it possible. It is possible in Swift, why not Android? Would be a nightmare to re-design the logic. If I had known this, I would  not choose Realm.

**Comment by missverma** [comment]
Are we going to have this functionality any time soon ? Any updates ?

**Comment by scorchgid** [comment]
Yeah also wanting this feature

**Comment by komuros** [comment]
It's not possible to use DataBinding Android feature without this issue or we duplicate all realmClass for wrap in BaseObservable class. Please add this in future version.

**Comment by Zhuinden** [comment]
@komuros actually, `BaseObservable` is a convenience class, although you *do* need to duplicate that across the RealmObjects to make it work.

See https://medium.com/@Zhuinden/realm-1-2-0-android-data-binding-1dc06822287f

**Comment by Zhuinden** [comment]
@cmelchior welp, hopefully once the Realm Cloud is up and running, hopefully the Core team will be able to focus on this kind of stuff - ObjectBox 1.4 seems to have introduced entity inheritance: http://objectbox.io/objectbox-1-4-property-queries-entity-inheritance/

**Comment by markiantorno** [comment]
Any updates on this task?

**Comment by omidraha** [comment]
> Any updates on this task?

Once in a while, 
Just get a promising email with this title of the issue, 
And then see nothing, just one message left to request it.

**Comment by noshaf** [comment]
When? 😐

**Comment by lordplagus02** [comment]
They should at least say something about it here, like "we're working on it" or "we're not working on it".

**Comment by markiantorno** [comment]
` It's Thursday morning,
Github emails an alert,
Maybe Realm is fixed?`

**Comment by markiantorno** [comment]
I am Mark's sense of disappointment.

**Comment by Zhuinden** [comment]
> Maybe Realm is fixed?

It's not broken :stuck_out_tongue: 

Anyways, if people want to bang some walls, they should do it on the Core issue instead: https://github.com/realm/realm-core/issues/778

**Comment by noshaf** [comment]
It's pretty broken without this... I have been continuing to use realm on my projects for some time in the hopes that this would be a priority and would have been added by now. IOS has had it for some time. Any ideas on when this might be implemented? Or if it will ever be implemented? It's causing my projects to be unnecessarily dirty and I would like to find an alternative to Realm if this is not going to be implemented anytime soon.

**Comment by missverma** [comment]
I, too, was waiting for them to add this functionality for a long time. But as it's not been added, I have started using Room with LiveData. It would have been very nice if only realm had this implemented.

**Comment by bmunkholm** [comment]
@noshaf @missverma Sorry for not addressing this sooner. Our highest priority has lately been to add synchronisation capabilities as that adds a whole lot of value and solves a major pain point for many apps and developers.

From an Android perspective, the thinking was that we wanted to do it properly and hence has been waiting for a proper solution in core. In contrast we have had some support for this for iOS with some limitations. What we are likely going to do in a "not too distant future" is to make the same support in Android, so we can at least solve some of the use cases, instead of waiting for the perfect solution. It's unfortunately not possible for me to be more precise about when we can get to it, but we are very soon going to plan it out.

In the meantime, if anyone is interested it would be possible to help contribute this feature at the same level as iOS has it :-) 

Thanks for the patience!

**Comment by Zhuinden** [comment]
Polymorphism is not supported in Swift either.

Personally, I think primitive list queries are more interesting than this; while it's not pretty, you can at least use single-table-inheritance without too much hacks.

Also, Kotlin lets you do

``` kotlin
interface Category {
     var name: String?
}
```

so you don't need to inherit getters/setters.

**Comment by ashughes** [comment]
I'm not familiar with what Realm iOS provides, but our needs are similar to what is mentioned in the OP. We're less interested in being able to share fields (although that would also be helpful) and more interested in being able to store a heterogeneous list of objects with the same base class (or interface). For example, with the classes defined in the OP, we might also have:

```java
public class Person extends RealmObject {
  private RealmList<Animal> pets;
}
```

This doesn't seem necessary in the above example, but for our use case the order of the objects in the list matters, which is why we can't just split up our data into multiple concrete lists.

@bmunkholm I assume that the "proper solution in core" is the only way to do this?

My understanding is that the only way to do this currently is to use a wrapper object, such as:

```java
public class AnimalWrapper extends RealmObject {
  // Only one of the following fields can be non-null
  private Dog dog;
  private Spider spider;

  public Animal getAnimal() {
    // return non-null animal
  }
}

public class Person extends RealmObject {
  private RealmList<AnimalWrapper> pets;
}
```

Where `Animal` is an interface that `Dog` and `Spider` implement.

**Comment by Zhuinden** [comment]
> We're less interested in being able to share fields (although that would also be helpful) and more interested in being able to store a heterogeneous list of objects with the same base class (or interface). 

That isn't supported by iOS either at this time.

-----

The only way to have "heterogenous collections" at this time is if you use the same class to model all child classes.

Have a field like

    private String type;

which specifies which child class it is.

``` java
public class MyParentObject extends RealmObject {
    @Index
    private String type;

    private String child1Property;

    private String child1OtherProperty;

    private String child2Property;

    private String child2OtherProperty;

    private String child2AnotherProperty;
}
```

That would allow storing items of different type. This is technically what is called `single-table inheritance`, and relational databases tend to use the same approach, except the ORM in-between allows the "syntactic sugar" of mapping them to different classes (child classes).

**Comment by ashughes** [comment]
> The only way to have "heterogenous collections" at this time is if you use the same class to model all child classes.

What about the object wrapper example I gave? I agree that it's not ideal, but this would still allow for model objects (and any logic) to be separated, rather than creating one giant model. Granted, it wouldn't allow for shared fields, but you could potentially get around that with some Kotlin trickery like you mentioned above. Still, I want proper inheritance.

My main point was just to say that on the surface, it doesn't sound like the limited support that @bmunkholm mentioned in the "not too distant future" would meet our needs. However, I'd like to hear more about what exactly that is (as I said, I'm not familiar with what Realm iOS does support).

**Comment by roberhofer** [comment]
Referenced in AHA: https://realmio.aha.io/epics/ROS-E-14

**Comment by VenomVendor** [comment]
Another user in the list waiting...

**Comment by markiantorno** [comment]
We're back on Room now. It works well and integrates no problem with the existing Android tools. I suggest everyone do the same.

**Comment by SergeyAppDroid** [comment]
@markiantorno I did not know anything about the Room. Does it have the same advantages as Realm?

**Comment by markiantorno** [comment]
@SergeyAppDroid 

It's basically a wrapper around SQLlite. However, it's built and maintained by Google, and a lot of the new Android Architecture components play very well with it. You'll also notice that the examples provided by Google on their site, all use Room. So if you are learning Android development, or referring to the existing documentation, it helps a lot.

Take a look here:
https://www.youtube.com/watch?v=SKWh4ckvFPM

https://developer.android.com/topic/libraries/architecture/room
https://developer.android.com/training/data-storage/room/defining-data

It's easy to set up and learn. Give it a try.

**Comment by SergeyAppDroid** [comment]
@markiantorno I'm a little upset that in the room you need to use SQL. And is there any support for List<Object> like Realm?

**Comment by bmunkholm** [comment]
@SergeyAppDroid @markiantorno Let's not pollute this issue with discussions about alternatives to Realm, we can have a pro/con discussion elsewhere :-), but let's stick to the issue here. Thanks!

This issue is certainly one that we intend to solve and although I really can't tell exactly when although I can promise that it won't take another year. I'm personally hoping for it as an xmas present.

**Comment by Zhuinden** [comment]
Personally I'm more hyped for primitive list queries.

But field sharing is nice too, technically if you can do it with a single table and discriminator fields by hand, then it ought to be possible in general. Although not a simple problem.

**Comment by barrysohl** [comment]
I would personally be happy with at least getting shared fields. All of our models have the same set of minimal fields (timestamp, created date, etc), and having to duplicate those fields, and the associated functionality, in every model is extremely cumbersome. Is there some other way to accomplish this currently in Realm that I am missing, or are we left to wait for some form of inheritance?

**Comment by MkazemAkhgary** [comment]
hours of implementation and realm punch in the face 😭

**Comment by hailhate** [comment]
> I would personally be happy with at least getting shared fields. All of our models have the same set of minimal fields (timestamp, created date, etc), and having to duplicate those fields, and the associated functionality, in every model is extremely cumbersome. Is there some other way to accomplish this currently in Realm that I am missing, or are we left to wait for some form of inheritance?

Composition helps a lot. Just put all your duplicatable fields in class and share it where it's needed

**Comment by Zhuinden** [comment]
>  Is there some other way to accomplish this currently in Realm that I am missing, or are we left to wait for some form of inheritance?

Simple way is to use Kotlin and your interfaces can define a `var` field and that works just fine 😄

**Comment by hardysim** [comment]
> Simple way is to use Kotlin and your interfaces can define a var field and that works just fine 😄

Another issue would be querying for the interface instead of the actual class. Is this possible?

E.g. I have an interface `A` with classes implementing it `B` and `C`.
Is it possible to "get all `A`"? Currently, I need to query for "all `B`" and "all `C`" and merge the result (casting to `A`).

**Comment by Zhuinden** [comment]
Okay, no, if that's what you need then the only thing you can do atm is you emulating polymorphism via single-table-inheritance-with-discriminator-field.

Meaning you have 1 big RealmObject that can possibly contain all fields of all "inheriting" objects and you also have a field called something like `type` or `objectType` or whatever which is most likely a String that is indexed and then you can query the data of a given type based on that

**Comment by omidraha** [comment]
Four more months passed. ;(

**Comment by SergeyAppDroid** [comment]
As a result, the problem will be solved? This is the main problem of realm. It makes use of it completely.

**Comment by Zhuinden** [comment]
I can give you a list of at least 3 things that are more important than this, considering for this you at least have a workaround at your disposal.

**Comment by omidraha** [comment]
@Zhuinden, Good idea,
Please publish roadmap list of realm, So we know where the priority of this issue is located.

**Comment by Zhuinden** [comment]
Don't look at me, I don't work for Realm and I can't see their priority list either :smile:

**Comment by Corical** [comment]
Bit of a joke, you follow proper polymorphic design and then you hit a stumbling block like this.

**Comment by roberhofer** [comment]
To give a brief update: 
- we are working on a improvement of the core DB structure (called Core6) that is a pre-cursor to enable inheritance / Polyphormism (and other features such as Cascading Deletes etc). 
- this work is actually now accelerated post-MongoDB acquisition (which is good news). 
- once that is in place then we will tackle the features/capabilities that have a dependency on this Core 6 improvement. 
- we heard you: we are actively working on it. 
- it will (still) take us months (though not years) to do it

**Comment by VenomVendor** [comment]
@roberhofer 
Polyphormism is definitely an improvement for devs. 
[Cascading Delete](https://github.com/realm/realm-core/issues/746), is the biggest issue, we have been struggling with it. Users are impacted with Orphan children in DB for ever.

Do you have any priority on these?

**Comment by jjimenez0611** [comment]
I'm waiting too... In my project I have a lot duplicate classes I need Inheritance...

**Comment by afsalashyana** [comment]
I am also eagerly waiting for this feature.

**Comment by muxammed** [comment]
I'm also ))

**Comment by malwinder-s** [comment]
Is there any update on this?

**Comment by Zhuinden** [comment]
You can check the road map where it is mentioned that it will happen one day.

**Comment by vernazza** [comment]
And here we are, at 2 years and a half since "I can promise that it won't take another year." (https://github.com/realm/realm-java/issues/761#issuecomment-405701707)
1 year and a half since "it will (still) take us months (though not years) to do it" (https://github.com/realm/realm-java/issues/761#issuecomment-499296327)

sigh :(

**Comment by ianpward** [comment]
hey all, product from Realm here, I am in the middle of writing the production description for this feature right now. As you might have guessed since almost no other databases exposes polymorphic inheritance, but rather leaves it up to the developer to build themselves, this is an incredibly difficult feature to build - but rest assured we are planning to build it. As mentioned in some of the comments above we had other high priority tasks we needed to complete first which would lay the foundation before inheritance - for instance we have released Cascading Deletes and are working on Maps and Mixed types right now which would be necessary for Polymorphism. Stay tuned.

**Comment by J-Rojas** [comment]
Not sure why people still choose to use Realm because of its heavy-handed restrictions.

ObjectBox is NoSQL and has supported inheritance for 2 years.

https://docs.objectbox.io/advanced/entity-inheritance

**Comment by tamimattafi** [comment]
> hey all, product from Realm here, I am in the middle of writing the production description for this feature right now. As you might have guessed since almost no other databases exposes polymorphic inheritance, but rather leaves it up to the developer to build themselves, this is an incredibly difficult feature to build - but rest assured we are planning to build it. As mentioned in some of the comments above we had other high priority tasks we needed to complete first which would lay the foundation before inheritance - for instance we have released Cascading Deletes and are working on Maps and Mixed types right now which would be necessary for Polymorphism. Stay tuned.

Any updates on this?

**Comment by ianpward** [comment]
Sorry I should hav posted an update to this ticket as soon as we had new information. We have finished the Initiative writeup and it has been approved internally - but Inheritance is just one part of this Initiative, there are still a  couple projects we need to implement before we start Inheritance. One of these migrations, you can imagine with Polymorphism that migrations will become a lot more complicated and we'd like to overhaul this feature so that users can start using Inheritance in production on Day 1. Another is we need to add some features to MongoDB Realm Schema so that our sync users can enjoy this feature too - for instance, we need to add the ability for $anyOf, $allOf, and $oneOf to our JSON schema support (which is how we plan to model Inheritance on the server-side). Polymorphism is actually utilized extensively in MongoDB so we expect sync clients to be some of our biggest users of Inheritance.  

We also need to wait for the Android team to have some free cycles. If you're not aware, the Android team has been heads down on building on a purely Kotlin SDK. 
https://github.com/realm/realm-kotlin

This comes with Multiplatform support out of the box and since we are building the SDK from the ground up we've taken this opportunity to address some longstanding functionality that has turned off some users. Most notably, the thread confinement of Realm Objects - this becomes an issue for developers who are used to programing with Coroutines natively.
This is why with this new Realm Kotlin SDK, we have removed this thread-confinement restriction and will have first-class support for Coroutines and Flows with built in APIs. It seems to be resonating with early users of our SDK:
https://github.com/realm/realm-kotlin/issues/377

So we have our hands full - I realize that this has been an issue for several years now but we've had a few "small" things happen in our business and technology space that we've had to prioritize - but we have now officially added it to our medium term roadmap and internally, we look forward to working on this and delivering it for our community.

**Comment by thipokch** [comment]
FYI Based on a simple test in Kotlin SDK, it seems that neither inheritance nor delegation pattern works. 
https://github.com/realm/realm-kotlin/issues/801

**Comment by afsalashyana** [comment]
Hi!

This is kinda nostalgic thread now. Any updates on this feature? Anything in the pipeline related to inheritance and polymorphism?

**Comment by tamimattafi** [comment]
I decided to unsubscribe to this thread. We had to come up with something to support polymorphism in our SQLite database (Using Room for android atm), and it worked. A little more boilerplate code, but it does the job.

**Comment by Zhuinden** [comment]
> Okay, no, if that's what you need then the only thing you can do atm is you emulating polymorphism via single-table-inheritance-with-discriminator-field.
> 
> Meaning you have 1 big RealmObject that can possibly contain all fields of all "inheriting" objects and you also have a field called something like `type` or `objectType` or whatever which is most likely a String that is indexed and then you can query the data of a given type based on that

This approach worked back in 2015 and still works to this day btw

**Comment by NizarETH** [comment]
Any updates on this feature?

**Comment by marchy** [comment]
@ianpward would you mind giving a "State of the Union" update on this issue?

[Your last such update](https://github.com/realm/realm-java/issues/761#issuecomment-907665073) in Summer 2021 was extremely helpful. We are looking to make some large architecture decisions based on inheritance support (in the Swift client for now, but Android just the same), so would be great to feel out a realistic timeframe for this – ie: is it imminent / expected this year VS not.

Cheers!

**Comment by vagrant1991** [comment]
can't believe this issue still open

**Comment by dhng22** [comment]
Almost 10 years for a feature, it seems like the team still has other priorities to concern. To me, realm is still an immature library to use. This issue is just one of those basic features that are being requested out there.

**Comment by longbombus** [comment]
I am glad that I encountered this problem at the start of using Realm. What is the benefit of NOSQL if the table fields are fixed as in SQL?
Now I'm looking for an alternative.

**Comment by Zhuinden** [comment]
@longbombus I'd be a little more concerned about the fact that the database is now fully abandoned by MongoDB (see the Atlas Device SDK deprecation docs)

Which is sad, technically it was pretty easy to do Inheritance, you did it manually with a discriminator field. Sure, it definitely wasn't first-party, but the workaround was doable. I dislike issues that have no workarounds. And stability concerns. Still, there was a lot of potential, but also a lot of potential lost to the 5 iterations of Realm Sync.
